include/dali/core/exec/tasking/sync.h

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_EXEC_TASKING_SYNC_H_
+#define DALI_CORE_EXEC_TASKING_SYNC_H_
+
+#include <algorithm>
+#include <memory>
+#include <mutex>
+#include "dali/core/small_vector.h"
+#include "dali/core/spinlock.h"
+
+namespace dali::tasking {
+
+class Scheduler;
+class Task;
+
+using SharedTask = std::shared_ptr<Task>;
+using WeakTask = std::weak_ptr<Task>;
+
+/** Represents any object in the tasking framework that a task can wait for.
+ *
+ * A waitable object may be passed to Task::Succeed. This causes the execution of the task to
+ * be deferred until all of its preceding conditions are ready.
+ *
+ * NOTE: The tasking framework implements "wait all" semantics - that is, a task acquires the
+ *       waitable objects only after all of them are available, thus mitigating deadlocks.
+ */
+class Waitable : public std::enable_shared_from_this<Waitable> {
+  friend class Scheduler;
+  friend class Task;
+
+ protected:
+  /** A list of tasks waiting for this waitable object. */
+  SmallVector<WeakTask, 8> waiting_;
+
+  /** Checks whether the Waitable is ready to be acquired. */
+  virtual bool IsAcquirable() const = 0;
+
+  /** @brief Changes the state of the object to acquired (if necessary) and returns whether the
+   *         operation was successful.
+   */
+  virtual bool AcquireImpl() = 0;
+
+  /** Informs the scheduler that this Waitable is signalled and some tasks may become unblocked. */
+  void Notify(Scheduler &sched);
+
+  /** Fast comparison of weak pointers without calling `lock`. */
+  static bool task_owner_equal(const WeakTask &w, const SharedTask &t) {
```

Comment:
```suggestion
  static bool task_owner_equal(const WeakTask &w, const WeakTask &t) {
```
Is there any value in relaxing the requirement that T is the SharedTask?